lib/rust/profiler/src/log.rs

```diff
@@ -0,0 +1,119 @@
+//! Data structure supporting limited interior mutability, to build up a collection.
+
+use std::cell;
+use std::mem;
+
+
+
+// ===========
+// === Log ===
+// ===========
+
+/// Data structure supporting limited interior mutability, to build up a collection.
+#[derive(Debug)]
+pub struct Log<T>(cell::UnsafeCell<Vec<T>>);
+
+#[allow(unsafe_code)]
+impl<T> Log<T> {
+    #[allow(clippy::new_without_default)]
+    /// Create a new, empty vec builder.
+    pub fn new() -> Self {
+        Self(cell::UnsafeCell::new(vec![]))
+    }
+
+    /// Push an element.
+    pub fn append(&self, element: T) {
+        // Note [Log Safety]
+        unsafe {
+            (&mut *self.0.get()).push(element);
+        }
+    }
+
+    /// Return (and consume) all elements pushed so far.
+    pub fn take_all(&self) -> Vec<T> {
+        // Note [Log Safety]
+        unsafe { mem::take(&mut *self.0.get()) }
+    }
+
+    /// The number of elements that are currently available.
+    pub fn len(&self) -> usize {
+        // Note [Log Safety]
+        unsafe { &*self.0.get() }.len()
+    }
+
+    /// Returns true if no elements are available.
+    pub fn is_empty(&self) -> bool {
+        // Note [Log Safety]
+        unsafe { &*self.0.get() }.is_empty()
+    }
+}
+
+#[allow(unsafe_code)]
+impl<T: Clone> Log<T> {
+    /// Return clones of all elements pushed so far.
+    pub fn clone_all(&self) -> Vec<T> {
+        // Note [Log Safety]
+        unsafe { &*self.0.get() }.to_vec()
+    }
+}
+
+// Note [Log Safety]
```

Comment:
I wonder whether this comment should not be in a more prominent place towards the beginning of the code using the unsafe functionality, or even as part of the struct docs.